internal/bigquery/batch/client.go

```diff
@@ -0,0 +1,185 @@
+)
+
+var (
+	autoDetectAndSchemaNotSupported = errors.New("BQ batch client: usage of autodetect being true and schema being passed is not supported")
+	autoDetectSchemaNotSupported    = errors.New("BQ batch client: autoDetect is only supported for JSON and CSV format")
+	csvOptionsNotAllowed            = errors.New("BQ batch client: csvOptions is only supported if the sourceFormat is CSV")
+	couldNotConvertReader           = errors.New("BQ batch client: could not convert data into io.Reader")
+	schemaRequired                  = errors.New("BQ batch client: schema is required if autoDetect is false")
+)
+
+// Client implements the standard/official BQ (cloud) Client,
+// using the regular insertAll API with retry logic added on top of that. By default
+// the workers will also batch its received rows rather than writing them one by one,
+// this can be disabled by setting the batchSize value to the value of 1.
+type Client struct {
+	client *bigquery.Client
+
+	dataSetID string
+	tableID   string
+
+	schema              *bigquery.Schema
+	sourceFormat        bigquery.DataFormat
+	autoDetect          bool
+	ignoreUnknownValues bool
+	csvOptions          *bigquery.CSVOptions
+	writeDisposition    bigquery.TableWriteDisposition
+
+	errors []*bigquery.Error
+}
+
+// NewClient creates a new Client.
+func NewClient(projectID, dataSetID, tableID string, ignoreUnknownValues, autoDetect bool, sourceFormat bigquery.DataFormat, writeDisposition *bigquery.TableWriteDisposition, schema *bigquery.Schema, csvOptions *bigquery.CSVOptions) (*Client, error) {
+	if projectID == "" {
+		return nil, fmt.Errorf("bq insertAll client creation: validate projectID: %w: missing", internal.InvalidParamErr)
+	}
+
+	// NOTE: we are using the background Context,
+	// as to ensure that we can always write to the client,
+	// even when the actual parent context is already done.
+	// This is a requirement given the streamer will batch its rows.
+	client, err := bigquery.NewClient(context.Background(), projectID)
+	if err != nil {
+		return nil, fmt.Errorf("BQ batch client: creation failed %w", err)
+	}
+
+	return newClient(
+		client, dataSetID, tableID,
+		ignoreUnknownValues, autoDetect,
+		sourceFormat, writeDisposition,
+		schema, csvOptions,
+	)
+}
+
+func newClient(client *bigquery.Client, dataSetID, tableID string, ignoreUnknownValues, autoDetect bool, sourceFormat bigquery.DataFormat, writeDisposition *bigquery.TableWriteDisposition, schema *bigquery.Schema, csvOptions *bigquery.CSVOptions) (*Client, error) {
+	if dataSetID == "" {
+		return nil, fmt.Errorf("BQ batch client: validate dataSetID: %w: missing", internal.InvalidParamErr)
+	}
+	if tableID == "" {
+		return nil, fmt.Errorf("BQ batch client: validate tableID: %w: missing", internal.InvalidParamErr)
+	}
+
+	// If autoDetect is passed while the format is not JSON or CSV, error as this is only supported for csv.
+	if autoDetect && (sourceFormat != bigquery.JSON && sourceFormat != bigquery.CSV) {
```

Comment:
The validation of the config should be moved to `StreamerConfig`, check its code there. I mean technically nothing wrong with doing it here, but so far I've done the configuration validation completely in StreamerConfig, such that I can assume that once it reaches the internal code that the config is completely fine and I can simply consume it without any checking.

Reply:
Indeed. I've moved it to its separate sanitise function